src/main/java/games/strategy/engine/framework/GameDataUtils.java

```diff
@@ -19,12 +24,13 @@ public static GameData cloneGameData(final GameData data) {
    * <strong>You should have the game data's read or write lock before calling this method</strong>
    */
   public static GameData cloneGameData(final GameData data, final boolean copyDelegates) {
-    try (ByteArrayOutputStream sink = new ByteArrayOutputStream(10000)) {
+    try {
+      final ByteArrayOutputStream sink = new ByteArrayOutputStream(10_000);
```

Comment:
I moved this out of the try-with-resources statement because it made it _appear_ that `sink` wasn't being flushed until after we used it to initialize `source`.  Obviously, there was no issue because neither `ByteArrayOutputStream` nor `ByteArrayInputStream` require an explicit flush or close.  Muh consistency, I guess.

Reply:
Worth adding a comment in the code that ByteArrayOutputStream does not need a `close()`?

Reply:
We create a new instance of `ByteArrayOutputStream` in 21 different locations.  Those occurrences are split between closing the stream and not closing it.

Would it be better if I submit a separate PR that adds the suggested comment to all occurrences where the instance is not being closed?  If desired, I could also modify the locations that do close it to no longer do so for consistency. 🤔

Reply:
At that point, we should check if we can use the more generic `OutputStream` type on the LHS : )

Personally, I would just close ByteArrayOutputStream anyways even if it does not matter.

Reply:
I just noticed the underscore in `10_000` - is that for readability? Does java now allow an underscore in numbers?

Reply:
A second comment, I suspect the 21 locations are probably borrowing/inspired by other locations, I would suspect a DRY violation.

Reply:
> Personally, I would just close ByteArrayOutputStream anyways even if it does not matter.

That was my philosophy, as well, until I ran into a problem similar to the code snippet we're commenting on.  That is, if you put a `ByteArrayOutputStream` in a try-with-resources statement, you can't call `toByteArray()` after the stream is closed due to the stream variable being out-of-scope.  And if you call `toByteArray()` before the end of the try-with-resources statement, you get the possible confusion for a code reader I described above.

In the past, I had used an explicit `close()` to try and resolve both forces:

```java
try {
  final ByteArrayOutputStream sink = new ByteArrayOutputStream(10000);
  GameDataManager.saveGame(sink, data, copyDelegates);
  sink.close();
  final ByteArrayInputStream source = new ByteArrayInputStream(sink.toByteArray());
  return GameDataManager.loadGame(source);
} catch (final IOException ex) {
  ClientLogger.logQuietly(ex);
  return null;
}
```

But, at least with Eclipse, the static analyzer then complains that "Resource 'sink' should be managed by try-with-resource". 😞 

Another long-term solution could be to encapsulate the detail of not having to close these types of streams into some utilities:

```java
@FunctionalInterface
interface OutputStreamConsumer {
  void accept(OutputStream sink) throws IOException;
}

static byte[] writeToMemory(final OutputStreamConsumer consumer) throws IOException {
  final ByteArrayOutputStream sink = new ByteArrayOutputStream(); // does not require close
  consumer.accept(sink);
  return sink.toByteArray();
}

@FunctionalInterface
interface InputStreamFunction<R> {
  R apply(InputStream source) throws IOException;
}

static <R> R readFromMemory(final byte[] bytes, final InputStreamFunction<R> function) throws IOException {
  final ByteArrayInputStream source = new ByteArrayInputStream(bytes); // does not require close
  return function.apply(source);
}
```

For example, the above code would become:

```java
try {
  final byte[] bytes = writeToMemory(sink -> GameDataManager.saveGame(sink, data, copyDelegates));
  return readFromMemory(bytes, source -> GameDataManager.loadGame(source));
} catch (final IOException e) {
  ClientLogger.logQuietly(e);
  return null;
}
```

Reply:
> I just noticed the underscore in `10_000` - is that for readability?

Yes.

> Does java now allow an underscore in numbers?

Yup, [underscores in numeric literals](http://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html) was introduced in Java 7 via Project Coin.

Reply:
Interesting, thanks for the reference

I would agree to extract the logic. I suspect that would also allow testing.

Reply:
Will do as soon as I get a chance.